### PR TITLE
Reduce overhead of `Waveform`

### PR DESCRIPTION
**Pull request by peppy**
Noticed that waveform points are up in the top 5 memory usages in editor (*game*-wide) so did a bit of refactoring. There's likely much more that can be improved here, but this was the low-hanging pass.

Before:

Parallels Desktop 2022-09-15 at 05 10 27

Parallels Desktop 2022-09-15 at 05 11 14

After:

Parallels Desktop 2022-09-15 at 05 07 45

Note that the counts here are a bit off (one more retained array in the "after" run) so savings are probably a touch better than displayed. Also the main goal was to switch away from `Point` being a `class`.

I also looked into reducing overall allocations during construction and resampling. The main fix for this was to reduce the `points_per_iteration` to a more sane number (it was previously allocating a very large buffer with no real benefit performance-wise). And then using `ArrayPool` for it (now that the size is beneath the `ArrayPool` allowance).

These results are probably more telling:

Before:

|       Method |  size |      Mean |     Error |    StdDev |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
|------------- |------ |----------:|----------:|----------:|---------:|---------:|---------:|----------:|
|   TestCreate |     ? | 99.921 ms | 1.2161 ms | 1.1375 ms | 800.0000 | 600.0000 | 400.0000 | 23,815 KB |
| TestResample |  1024 |  3.468 ms | 0.0175 ms | 0.0164 ms |   7.8125 |        - |        - |     98 KB |
| TestResample | 32768 |  6.994 ms | 0.1383 ms | 0.1480 ms | 257.8125 | 140.6250 | 109.3750 |  3,074 KB |

After:

|       Method |  size |      Mean |     Error |    StdDev |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
|------------- |------ |----------:|----------:|----------:|---------:|---------:|---------:|----------:|
|   TestCreate |     ? | 91.525 ms | 0.8581 ms | 0.8026 ms | 333.3333 | 166.6667 | 166.6667 |  3,614 KB |
| TestResample |  1024 |  2.323 ms | 0.0098 ms | 0.0082 ms |   3.9063 |        - |        - |     58 KB |
| TestResample | 32768 |  3.421 ms | 0.0214 ms | 0.0200 ms | 285.1563 | 257.8125 | 195.3125 |  1,795 KB |

### CODE REVIEW

**Comment by smoogipoo** [comment]
Clicking the "NullWaveform" button in the test scene crashes with
```
[runtime] 2022-09-15 09:42:18 [error]: System.IndexOutOfRangeException: Index was outside the bounds of the array.
[runtime] 2022-09-15 09:42:18 [error]: at osu.Framework.Graphics.Audio.WaveformGraph.WaveformDrawNode.Draw(IRenderer renderer) in /home/smgi/Repos/osu-framework/osu.Framework/Graphics/Audio/WaveformGraph.cs:line 382
```
Not sure why it's not failing on CI - pretty scary.

**Comment by peppy** [comment]
> Not sure why it's not failing on CI - pretty scary.

Probably this:

https://github.com/ppy/osu-framework/blob/66d72918a1fece087665c643d0cb3194790973c1/osu.Framework/Audio/Track/Waveform.cs#L87

**Comment by peppy** [comment]
Definitely worth the change:

|       Method |  size |      Mean |     Error |    StdDev |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
|------------- |------ |----------:|----------:|----------:|---------:|---------:|---------:|----------:|
|   TestCreate |     ? | 81.630 ms | 0.4930 ms | 0.4612 ms | 142.8571 | 142.8571 | 142.8571 |  1,317 KB |
| TestResample |  1024 |  2.057 ms | 0.0041 ms | 0.0038 ms |        - |        - |        - |     22 KB |
| TestResample | 32768 |  2.283 ms | 0.0042 ms | 0.0035 ms | 164.0625 | 164.0625 | 164.0625 |    643 KB |

File: osu.Framework/Audio/Track/Waveform.cs [hunk 1]
```diff
@@ -116,24 +124,30 @@ public Waveform([CanBeNull] Stream data)
...
-                            var point = new Point(info.Channels);
+                            var point = new Point();
 
                             for (int j = i; j < i + samplesPerPoint; j += info.Channels)
                             {
                                 // Find the maximum amplitude for each channel in the point
-                                for (int c = 0; c < info.Channels; c++)
-                                    point.Amplitude[c] = Math.Max(point.Amplitude[c], Math.Abs(sampleBuffer[j + c]));
+                                point.AmplitudeLeft = Math.Max(point.AmplitudeLeft, Math.Abs(sampleBuffer[j]));
+                                point.AmplitudeRight = Math.Max(point.AmplitudeLeft, Math.Abs(sampleBuffer[j + secondChannelIndex]));
```
Comment (smoogipoo) [inline osu.Framework/Audio/Track/Waveform.cs:143]:
```suggestion
                                point.AmplitudeRight = Math.Max(point.AmplitudeRight, Math.Abs(sampleBuffer[j + secondChannelIndex]));
```

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**